### ISSUE TITLE
bug: Incompatibility between vdirsyncer and khard modules

### ISSUE THREAD

**Issue by DamienCassou** [nix-community/home-manager#4531]
### Are you following the right branch?

- [X] My Nixpkgs and Home Manager versions are in sync

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Issue description

I have enabled both khard and vdirsyncer programs. The idea is that vdirsyncer fetches&uploads the contacts that are listed&edited with khard.

It seems that it isn't possible to have both vdirsyncer and khard configured together with home-manager because:
1. vdirsyncer creates a subfolder within `accounts.contact.accounts.<name>.local.path` and puts the vcf files in the subfolder.
2. khard reads the vcf files from `accounts.contact.accounts.<name>.local.path` directly and doesn't look recursively inside the subfolders.
As a result, khard finds no vcf files and thus lists no contacts.

Said differently, here is part of my vdirsyncer configuration:

```ini
[storage contacts_nico_local]
fileext = ".vcf"
path = "~/configuration/contacts"
```

and here is part of my *working* hand-written khard configuration:

```ini
[addressbooks]
[[contacts]]
path = ~/configuration/contacts/172ce...
```

As you can see, the configuration file references the subfolder `172ce.../` instead of directly the path configured for vdirsyncer.

Here is an excerpt of what home-manager generates for khard:

```ini
[addressbooks]
[[nico]]
path = ~/configuration/contacts
```

This doesn't work because of the missing subfolder.

Here is the home-manager configuration that I cannot use:

```nix
{ pkgs, lib, ... }:

{
  programs.vdirsyncer = { enable = true; };

  programs.khard = {
    enable = true;
    settings = {
      ...
    };
  };

  accounts.contact.accounts = {
    nico = {
      vdirsyncer = {
        enable = true;
        ...
      };
      khard = { enable = true; };
      local = {
        path = "~/configuration/contacts";
        ...
      };
      remote = { ... };
    };
  };
}
```

### Maintainer CC

@olmokramer @antonmosich @teto @ayyess @loewenheim @jorsn 

### System information

```markdown
- system: `"x86_64-linux"`
 - host os: `Linux 6.5.5-200.fc38.x86_64, Fedora Linux, 38 (Workstation Edition), nobuild`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.17.0`
 - channels(root): `"nixpkgs"`
 - channels(temp): `"home-manager, nixpxkgs"`
 - channels(cassou): `"home-manager-22.11.tar.gz, nixpkgs"`
 - nixpkgs: `/home/cassou/Documents/projects/nix-system/nixpkgs`
```

**Comment by antonmosich** [comment]
Thanks for the detailed report. I have a very similar config working correctly on my machine, so I'll need more details to figure out where the problem is. With my config, vdirsyncer puts the vcf files in `local.path` directly, without having an "extra" directory. My config has
```
[storage contacts_nico_local]
fileext = ".vcf"
path = "~/Contacts/Privat"
post_hook = "..."
type = "filesystem"
```
Maybe your `type` is different? Could you check that?

**Comment by DamienCassou** [comment]
They look very similar to me:

```ini
[storage contacts_nico_local]
fileext = ".vcf"
path = "~/configuration/contacts"
post_hook = "..."
type = "filesystem"

[storage contacts_nico_remote]
password.fetch = ["command", "..."]
type = "carddav"
url = "https://<server>"
username = "damien"
```

I tried to change `url` from "https://<server>" to "https://<server>/damien/172ce...". Both work fine but both make vdirsyncer create a subdirectory below `~/configuration/contacts`.

- vdirsyncer, version 0.19.2
- I'm using [Radicale](https://radicale.org) as CardDAV Server

**Comment by DamienCassou** [comment]
I understand what is happening and found a workaround. In vdirsyncer's configuration, I had:

```ini
[pair contacts_nico]
a = "contacts_nico_remote"
b = "contacts_nico_local"
collections = ["from b", "from a"]
```

According to [the documentation of `path`](https://vdirsyncer.pimutils.org/en/stable/config.html#local), setting `collections` this way makes the path be a list of collections (i.e., a directory with subdirectories) instead of just a collection (i.e., a directory with vcf files). This is further explained in [the tutorial's section about collections](https://vdirsyncer.pimutils.org/en/stable/tutorial.html#more-information-about-collections).

The workaround for me is to set `collections` to `null` and `url` to `"https://<server>/damien/172ce..."` instead of just `"https://<server>"`.

Unfortunately, the [documentation says](https://vdirsyncer.pimutils.org/en/stable/tutorial.html#more-information-about-collections):

> Note that not all storages support the null-collection, for example [google_contacts](https://vdirsyncer.pimutils.org/en/stable/config.html#storage-google_contacts) and [google_calendar](https://vdirsyncer.pimutils.org/en/stable/config.html#storage-google_calendar) don’t.

This means home-manager users of these 2 storages won't be able to get both khard and vdirsyncer to work together.

**Comment by olmokramer** [comment]
Thanks for the report @DamienCassou and for further digging into this! I don't set the `collections` key in my config so that's why it slipped by me. I also use a radicale server for my contacts and calendar, so I'll take a deeper look during the weekend to see if we can support both use cases on that configuration.

**Comment by zoglesby** [comment]
I have been running into the same issue with fastmail's carddav.

I have the following in my collections, that should allow for a programmatic setup of paths.

`collections = [
   ["main"` "Default" "main" ]
   ["bd" "UUID" "birthday" ]
];`

The first item in the collection is a name that vdirsyncer uses internally, the second the name on the `a` side in our case the remote. The final item is the name of the subdirectory on the local side.

Using this you should be able to add entries into `[addressbooks]` for each collection rather than the account top level.

**Comment by olmokramer** [comment]
@zoglesby I'm sorry, life got really busy in the meantime and I haven't been able to take a look yet. Will try to make some time the upcoming week :)

**Comment by jorsn** [comment]
I had this problem already back when I created my own khard modules. My workaround was to hardcode the sub-accounts (collections) for khard:

```nix
let
  # ...
  localVcf = {
    type = "filesystem";
    fileExt = ".vcf";
  };
  khardAbook = { khard.enable = true; local = localVcf; };
in
  accounts.contact = {
    basePath = "contacts";
    accounts."jorsn/contacts" = khardAbook;
    accounts."jorsn/Collected" = khardAbook;

    accounts.jorsn = {
      # khard.enable = true; # khard cannot auto-discover sub calendars,
      # thus see above.

      vdirsyncer = defaultVdirsyncer;

      local = localVcf;

      remote = ncRemote "carddav";
    };
  };
```

As @zoglesby [proposed](https://github.com/nix-community/home-manager/issues/4531#issuecomment-1799775880), one should be able to auto-generate the accounts if the collections are not auto-discovered, but hardcoded in the vdirsyncer config.
Khard does not auto-discover the accounts, so they have to be hardcoded *somewhere* anyway.

**Comment by olmokramer** [comment]
Ok I've finally had some time to take a look at this and as @jorsn mentions it seems impossible to have both auto-discovery of collections and khard automagically use those collections. Khard needs to be configured with a path to a directory directly containing the `.vcf` files and so the collection name needs to be hardcoded _somewhere. So it will still be a pain to use with google calendar/contacts.

@zoglesby it would indeed be possible to generate addressbooks from that configuration. However, I've never used vdirsyncer collections, so I don't know what addressbooks configurations you would expect to be generated for those collections. Could you provide an example of the expected configuration? Also, it seems to be allowed to provide a single string for a collection instead of a triple of strings. What should be generated in that case?

**Comment by jorsn** [comment]
> Khard needs to be configured with a path to a directory directly containing the .vcf files and so the collection name needs to be hardcoded _somewhere. So it will still be a pain to use with google calendar/contacts.

...which only means that the proper place to fix this is in khard, not in home-manager.

**Comment by olmokramer** [comment]
> ...which only means that the proper place to fix this is in khard, not in home-manager.

Well, there's two separate problems. One is that we currently cannot support dynamic discovery for khard, which indeed can only be fixed in khard itself. So if people need that I would suggest opening an issue on the khard repo.

But there's also the issue that the khard module as is is lacking support for even hardcoded collections. So I'd still like to know how people expect that to work so that I can either implement it or review a PR that does.

**Comment by octvs** [comment]
In case anybody is still interested: 

#5220 didn't spring from the discussion here rather it was just my local fix,
that lets the user to define a hardcoded path, getting upstreamed. Albeit it
doesn't address aforementioned auto-discovery, it allows a way out. 

In case it shouldn't have closed this issue, forgive me for the inconvenience.

**Comment by octvs** [comment]
While setting up several contact sets from the same nextcloud account, this
became relevant again.

For anyone who wants to achieve a working setup with the current module, the
following works for three calendars `calendar-{a,b,c}` on the same account.

```nix
accounts.contact.accounts = rec {
  main = {
    local = {
      type = "filesystem";
      path = "<some_path>/contacts/";
      fileExt = ".vcf";
    };
    # ...
    vdirsyncer = {
      enable = true;
      collections = ["calender-a" "calender-b" "calender-c"];
      # ...
    };
    khard = {
      enable = true;
      defaultCollection = "calender-a";
    };
  };
  # You have two options to have the extra address book
  # Either define it as the default collection with the same `local`
  nextcloud-calendar-b = {
    local = main.local;
    khard = {
      enable = true;
      defaultCollection = "calendar-b";
    };
  };
  # Or hardcode it to the local.path 
  nextcloud-calendar-c = {
    local.path = "${main.local.path}/calendar-c";
    khard.enable = true;
  };
};
```

I will give it a try to provide another patch that achieves the same without
needing to define an extra account.

**Comment by vbrandl** [comment]
I'm not sure if this is related or if I should start a new issue, but I'll start by asking here. I want to sync my contacts from my Nextcloud instance. I wrote the following config:

```nix
{ pkgs, ... }: {
  programs.khard.enable = true;

  services.vdirsyncer.enable = true;

  accounts.contact = {
    basePath = "contacts";
    accounts = {
      nextcloud_contacts = {
        khard = {
          enable = true;
        };
        remote = {
          userName = "<snip>";
          passwordCommand = [...];
          type = "carddav";
          url = "<snip>";
        };
        vdirsyncer = {
          enable = true;
          collections = ["from a" "from b"];
        };
      };
    };
  };
}
```

When I try to build this configuration, I get this error:

```
error:
       … while calling the 'head' builtin
         at /nix/store/97hykhxj3miqqhfnf55b1rphiwcpxlcr-source/lib/attrsets.nix:1534:13:
         1533|           if length values == 1 || pred here (elemAt values 1) (head values) then
         1534|             head values
             |             ^
         1535|           else

       … while evaluating the attribute 'value'
         at /nix/store/97hykhxj3miqqhfnf55b1rphiwcpxlcr-source/lib/modules.nix:991:7:
          990|     // {
          991|       value = addErrorContext "while evaluating the option `${showOption loc}':" value;
             |       ^
          992|       inherit (res.defsFinal') highestPrio;

       … while evaluating the option `system.build.toplevel':

       … while evaluating definitions from `/nix/store/97hykhxj3miqqhfnf55b1rphiwcpxlcr-source/nixos/modules/system/activation/top-level.nix':

       … while evaluating the option `assertions':

       … while evaluating definitions from `/nix/store/avba4k04vxmknff9mjchmwyvr3vpzccz-source/nixos/common.nix':

       … while evaluating the option `home-manager.users.me.assertions':

       … while evaluating definitions from `/nix/store/avba4k04vxmknff9mjchmwyvr3vpzccz-source/modules/programs/vdirsyncer.nix':

       (stack trace truncated; use '--show-trace' to show the full, detailed trace)

       error: expected a set but found null: null
```

Are those related or should I open a new issue?

**Comment by octvs** [comment]
_I think your issue is already handled by a PR. But here is context
nevertheless, for reference._ 

> I'm not sure if this is related or if I should start a new issue 

This is not related to this issue; [as summarized above][prev-comment] the
issue was about two separate problems:

- Khard not supporting dynamic discovery 
  - This should be fixed upstream. There is already [an issue][khard-issue].
- Khard home-manager module not supporting hardcoded collections.
  - This was addressed by #5220.

[prev-comment]: https://github.com/nix-community/home-manager/issues/4531#issuecomment-1879830325
[khard-issue]: https://github.com/lucc/khard/issues/203

**Comment by vbrandl** [comment]
Yeah, in the end I realized that my problem was not related and I created the PR to fix my problem. Anyway thanks for the context.
